lexicon/providers/godaddy.py

```diff
@@ -0,0 +1,131 @@
+
+    def authenticate(self):
+        domain = self.options.get('domain')
+
+        result = self._get('/domains/{0}'.format(domain))
+        self.domain_id = result['domainId']
+
+    def create_record(self, type, name, content):
+        domain = self.options.get('domain')
+        ttl = self.options.get('ttl')
+
+        data = {'data': content}
+        if ttl:
+            data['ttl'] = ttl
+
+        self._put('/domains/{0}/records/{1}/{2}'
+                  .format(domain, type, self._relative_name(name)), [data])
+
+        LOGGER.debug('create_record: %s %s %s', type, name, content)
+
+        return True
+
+    def list_records(self, type=None, name=None, content=None):
+        domain = self.options.get('domain')
+        records = []
+
+        url = '/domains/{0}/records'.format(domain)
+        if type:
+            url += '/{0}'.format(type)
+        if name:
+            url += '/{0}'.format(self._relative_name(name))
+
+        raws = self._get(url)
+
+        for raw in raws:
+            records.append({
+                'type': raw['type'],
+                'name': self._full_name(raw['name']),
+                'ttl': raw['ttl'],
+                'content': raw['data']
+            })
+
+        if content:
+            records = [record for record in records if record['data'].lower() == content.lower()]
+
+        LOGGER.debug('list_records: %s', records)
+
+        return records
+
+    def update_record(self, identifier, type=None, name=None, content=None):
+        # With GoDaddy API, creating a record for given type and name is the same
+        #   than updating the record.
+        return self.create_record(type, name, content)
+
+    def delete_record(self, identifier=None, type=None, name=None, content=None):
+        domain = self.options.get('domain')
+
+        if not type:
+            raise Exception('ERROR: type must be setted')
+        if not name:
+            raise Exception('ERROR: name must be setted')
+        if not content:
+            raise Exception('ERROR: content must be setted')
+
+        # OK some explanations need to be done here.
+        # GoDaddy DNS API does not provide a direct way to delete a record (weird).
+        # However it provides a way to get and update all records of a zone.
+        # So :
+        #  - we get all the records,
+        #  - we filter the array to remove the record to be deleted,
+        #  - then we push back the filtered array to set the zone without the record to be deleted.
+        # And yes, we could limit the operation to a given type record (eg. TXT, there is an URL
+        #   for that), but GoDaddy refuses to push back an empty set of a given type (yep, you
+        #   cannot remove all your TXT with this URL, ultra weird).
+        # It is likely to happen during a DNS challenge, as all TXT should be removed at the end.
+        # So operating on all the zone avoid empty sets (there will always at least NS entries).
```

Comment:
Yep really weird. Even more when you consider their approach about "merging" insert and update functionalities which is in fact pretty relevant.